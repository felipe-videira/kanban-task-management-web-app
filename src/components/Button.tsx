import { bool, func, InferProps } from "prop-types";
import styled from "styled-components/macro";

const propTypes = {
  outlined: bool,
  small: bool,
  icon: bool,
  onClick: func.isRequired,
};

const Button = styled.button.attrs(() => ({
  type: "button",
}))<InferProps<typeof propTypes>>`
  width: 100%;
  font-family: inherit;
  border-radius: 7px;
  text-transform: uppercase;
  letter-spacing: 3px;
  font-size: 1rem;
  cursor: pointer;
  padding: ${(props) => (props.small ? "10px" : "14px")};

  ${(props) =>
    props.outlined
      ? `
        background: transparent;
        border: 2px solid ${props.theme.contrast.medium};
        color: ${props.theme.primary};`
      : `
        background: ${props.theme.primary};
        border: none;
        color: ${props.theme.secondary};`}

  ${(props) =>
    props.icon
      ? `
        border: none;
        background: none;
        border-radius: 50%;
        cursor: pointer;
        font-size: 2rem;
        color: ${props.theme.contrast.low};
        display: flex;
        align-items: center;
        justify-content: center;
        letter-spacing: 0;
        line-height: 0;
        padding: 0;
        svg {
          fill: ${props.theme.contrast.low};
        }`
      : ""};
`;

Button.propTypes = propTypes;

export default Button;
