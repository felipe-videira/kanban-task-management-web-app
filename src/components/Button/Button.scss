@import "../../styles/variables";
@import "../../styles/mixins";
@import "../../styles/functions";

.base-button {
  border: none;
  font-family: "Plus Jakarta Sans", sans-serif;
  font-weight: 700;
  padding: 0.75rem 1.5rem;
  font-size: 0.75rem;
  border-radius: 50px;
  cursor: pointer;

  &--block {
    width: 100%;
    margin: 0 0 1rem;
  }
}

@include themify($themes) {
  .base-button {
    &--primary {
      background-color: themed("primary");
      color: themed("primary-btn-text");

      &:hover {
        background-color: themed("primary-hover");
      }
    }

    &--secondary {
      background-color: themed("secondary-btn-bg");
      color: themed("primary");

      &:hover {
        background-color: themed("secondary-btn-hover");
      }
    }

    &--danger {
      background-color: themed("danger-btn-bg");
      color: themed("danger-btn-text");

      &:hover {
        background-color: themed("danger-btn-hover");
      }
    }
  }
}
