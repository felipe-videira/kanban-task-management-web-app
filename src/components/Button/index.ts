import { bool, func, InferProps } from "prop-types";
import styled from "styled-components/macro";

const propTypes = {
  outlined: bool,
  small: bool,
  icon: bool,
  onClick: func.isRequired,
};

const Button = styled.button.attrs(() => ({
  type: "button",
}))<InferProps<typeof propTypes>>`
  ${(props) =>
    props.outlined
      ? `
    background: transparent;
    border: 2px solid ${props.theme.contrast.medium};
    color: ${props.theme.primary};`
      : `
    background: ${props.theme.primary};
    border: none;
    color: ${props.theme.secondary};`}

  ${(props) =>
    props.small
      ? `
      padding: 10px;
      max-width: 150px;`
      : `
      padding: 14px;
      max-width: 350px;`}
      
  font-size: 1rem;
  width: 100%;
  font-family: inherit;
  border-radius: 7px;
  text-transform: uppercase;
  letter-spacing: 2px;
  cursor: pointer;
  transition: transform 300ms ease;

  &:hover {
    ${(props) => (props.icon ? "" : "transform: scale(1.05);")}
  }

  ${(props) =>
    props.icon
      ? `
    border: none;
    background: none;
    width: 50px;
    height: 50px;
    border-radius: 50%;
    cursor: pointer;
    font-size: 2rem;
    color: ${props.theme.contrast.low};
    display: flex;
    align-items: center;
    justify-content: center;
    letter-spacing: 0;
    line-height: 0;
    padding: 0;

    svg {
      fill: ${props.theme.contrast.low};
      height: inherit;
      width: inherit;
    }
    
    `
      : ""}
`;

Button.propTypes = propTypes;

export default Button;
