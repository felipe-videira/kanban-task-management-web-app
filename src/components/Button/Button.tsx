import "./Button.scss";
import { bool, func, node, string } from "prop-types";
import { ReactNode } from "react";

function Button({
  children,
  type,
  onClick,
  className,
  block,
  primary,
  secondary,
  danger,
}: {
  children: ReactNode;
  type: "button" | "submit" | "reset" | undefined;
  onClick: () => void;
  className: string;
  block: boolean;
  primary: boolean;
  secondary: boolean;
  danger: boolean;
}) {
  return (
    <button
      type={type}
      onClick={onClick}
      className={`base-button ${className} ${block ? "base-button--block" : ""}
      ${primary ? "base-button--primary" : ""}
      ${secondary ? "base-button--secondary" : ""}
      ${danger ? "base-button--danger" : ""}
      `}
    >
      {children}
    </button>
  );
}

Button.propTypes = {
  children: node.isRequired,
  onClick: func,
  type: string,
  className: string,
  block: bool,
  primary: bool,
  secondary: bool,
  danger: bool,
};

Button.defaultProps = {
  type: "button",
  onClick: null,
  className: "",
  block: false,
  primary: true,
  secondary: false,
  danger: false,
};

export default Button;
