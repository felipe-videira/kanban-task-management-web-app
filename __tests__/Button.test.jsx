import "jsdom-global/register";
import "jest-styled-components";

import React from "react";
import { render, shallow } from "enzyme";
import Button from "../src/components/Button";

describe("Button", () => {
  it('should render correctly with "outlined" true', () => {
    const wrapper = render(<Button outlined onClick={() => {}} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with "outlined" false', () => {
    const wrapper = render(<Button outlined={false} onClick={() => {}} />);

    expect(wrapper).toMatchSnapshot();
  });

  it("should call click function", () => {
    const onClick = jest.fn();

    const wrapper = shallow(<Button outlined onClick={onClick} />);

    wrapper.simulate("click");

    expect(onClick).toHaveBeenCalled();
  });
});
