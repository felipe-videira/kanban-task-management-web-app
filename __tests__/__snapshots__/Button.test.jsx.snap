// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly with "outlined" false 1`] = `
.c0 {
  background: #fff;
  border: none;
  color: #1f3756;
  border: none;
  width: 100%;
  font-size: 16px;
  font-family: inherit;
  padding: 14px;
  border-radius: 7px;
  text-transform: uppercase;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  cursor: pointer;
}

<button
  class="c0"
  type="button"
/>
`;

exports[`Button should render correctly with "outlined" true 1`] = `
.c0 {
  background: transparent;
  border: 1px solid #fff;
  color: #fff;
  border: none;
  width: 100%;
  font-size: 16px;
  font-family: inherit;
  padding: 14px;
  border-radius: 7px;
  text-transform: uppercase;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  cursor: pointer;
}

<button
  class="c0"
  type="button"
/>
`;
